asllib/Typing.ml

```diff
@@ -3625,81 +3625,133 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
-          let res0 = from_stmt body in
-          let res1 =
-            match otherwise with
-            | None -> res0
-            | Some s -> join (from_stmt s) res0
-          in
-          List.fold_left
-            (fun res (_, _, s) -> join res (from_stmt s))
-            res1 catchers
+    let union configs1 configs2 = normalize (configs1 @ configs2)
 
-    (* End *)
+    let pp_abs_config = function
+      | Abs_Throwing -> "Abs_Throwing"
+      | Abs_Returning -> "Abs_Returning"
+      | Abs_Continuing -> "Abs_Continuing"
+      | Abs_DynError -> "Abs_DynError"
 
-    (** [check_stmt_interrupts name body] checks that the function named [name]
-        with the statement body [body] either: returns a value, throws an
-        exception, or calls [Unreachable()].
-        It executes only when [C.control_flow_analysis] is [true]. *)
-    let check_stmt_returns_or_throws name s () =
-      if C.control_flow_analysis then
-        match from_stmt s with
-        | AssertedNotInterrupt | Interrupt -> ()
-        | MayNotInterrupt -> fatal_from ~loc:s (Error.NonReturningFunction name)
+    let pp_abs_configs fmt (configs : abs_config list) =
+      let pp_config _fmt c = Format.eprintf "%s" (pp_abs_config c) in
+      Format.pp_print_list
+        ~pp_sep:(fun fmt () -> Format.fprintf fmt "; ")
+        pp_config fmt configs
+
+    (** [approx_stmt senv s] returns a superset of the abstract configurations
+        that evaluating [s] with any environment consisting of [senv] would yield.
+        The analysis conservatively assumes that evaluating any expression
+        results in either a value or a thrown exception.
+    *)
+    let rec approx_stmt senv s =
+      let configs =
+        match s.desc with
+        | S_Pass | S_Decl _ | S_Assign _ | S_Assert _ | S_Print _ | S_Pragma _
+          ->
+            [ Abs_Continuing; Abs_Throwing; Abs_DynError ]
+        | S_Unreachable -> [ Abs_DynError ]
+        | S_Call call ->
+            let f, _ = IMap.find call.name senv.global.subprograms in
+            if ASTUtils.is_noreturn f then [ Abs_Throwing; Abs_DynError ]
+            else [ Abs_Continuing; Abs_Throwing; Abs_DynError ]
+        | S_Return _ -> [ Abs_Returning; Abs_Throwing; Abs_DynError ]
+        | S_Throw _ -> [ Abs_Throwing; Abs_DynError ]
+        | S_Seq (s1, s2) ->
+            let configs1 = approx_stmt senv s1 in
+            let configs2 = approx_stmt senv s2 in
+            List.map
+              (fun c1 ->
+                match c1 with Abs_Continuing -> configs2 | _ -> [ c1 ])
+              configs1
+            |> List.flatten
+        | S_Cond (_, s1, s2) ->
+            let configs1 = approx_stmt senv s1 in
+            let configs2 = approx_stmt senv s2 in
+            union configs1 configs2
+        | S_Repeat (body, _, _) | S_For { body } | S_While (_, _, body) ->
+            let configs_body = approx_stmt senv body in
+            union configs_body [ Abs_Throwing; Abs_DynError ]
```

Comment:
nit: I wonder if it is worth factoring out something like `expr_configurations = [ Abs_Throwing; Abs_DynError ]` and reusing this when expressions are involved